### PR TITLE
WIP Modernize to use convert, porting over mexjulia changes

### PR DESCRIPTION
**Pull request by musm**

### CODE REVIEW

File: src/mxarray.jl [hunk 1]
```diff
@@ -15,12 +15,12 @@ type MxArray
 end
 MxArray(p::Ptr{Void}) = MxArray(p, true)
 
-mxarray(mx::MxArray) = mx
+MxArray(mx::MxArray) = mx
```
Comment (yuyichao) [inline src/mxarray.jl]: Is this needed?

File: src/mxarray.jl [hunk 2]
```diff
@@ -157,13 +158,16 @@ const _mx_is_complex = mxfunc(:mxIsComplex)
...
-const _mx_is_cell    = mxfunc(:mxIsCell) 
+const _mx_is_cell    = mxfunc(:mxIsCell)
 
 
 # getting simple attributes
 
 macro mxget_attr(fun, ret)
-    :( ccall($(fun)::Ptr{Void}, $(ret), (Ptr{Void},), mx.ptr) )
+    quote
+        @assert mx.ptr::Ptr{Void} != C_NULL
```
Comment (yuyichao) [inline src/mxarray.jl]: I don't know what other places in the package uses but this shouldn't be a `@assert`, which should only be used for internal consistency check and shouldn't be anything the user can trigger. I usually use `UndefRefError` for this kind of error. This should also use `unsafe_convert` instead but that can be another PR.

File: src/mxarray.jl [hunk 3]
```diff
@@ -652,15 +647,22 @@ function jdict(mx::MxArray)
...
-function jvariable(mx::MxArray)
+function Function(mx::MxArray)
+    if classid(mx) != mxFUNCTION_CLASS || nelems(mx) != 1
+        throw(ArgumentError("Function only applies to a single function handle."))
+    end
+    return (args...) -> mx(args...)
+end
+
+
+function Any(mx::MxArray)
```
Comment (yuyichao) [inline src/mxarray.jl]: As @simonster said, this is very wrong.

File: src/mxarray.jl [hunk 4]
```diff
@@ -670,24 +672,26 @@ function jvariable(mx::MxArray)
...
 
-jvariable(mx::MxArray, ty::Type{Array})  = jarray(mx)
-jvariable(mx::MxArray, ty::Type{Vector}) = jvector(mx)
-jvariable(mx::MxArray, ty::Type{Matrix}) = jmatrix(mx)
-jvariable(mx::MxArray, ty::Type{Number}) = jscalar(mx)::Number
-jvariable(mx::MxArray, ty::Type{String}) = jstring(mx)::String
-jvariable(mx::MxArray, ty::Type{Dict}) = jdict(mx)
-jvariable(mx::MxArray, ty::Type{SparseMatrixCSC}) = jsparse(mx)
-
+convert(::Type{Array}, mx::MxArray)  = jarray(mx)
```
Comment (yuyichao) [inline src/mxarray.jl]: This will implicitly convert `MxArray` to these types and I don't think all of these should happen.

File: src/mxarray.jl [hunk 5]
```diff
@@ -670,24 +672,26 @@ function jvariable(mx::MxArray)
...
+convert(::Type{Vector}, mx::MxArray) = jvector(mx)
+convert(::Type{Matrix}, mx::MxArray) = jmatrix(mx)
+convert(::Type{Number}, mx::MxArray) = jscalar(mx)::Number
```
Comment (yuyichao) [inline src/mxarray.jl]: Did you only changed some of the signatures or is the RHS here missing from the change?

File: src/mxarray.jl [hunk 6]
```diff
@@ -670,24 +672,26 @@ function jvariable(mx::MxArray)
...
+convert(::Type{String}, mx::MxArray) = String(mx)::String
+convert(::Type{SparseMatrixCSC}, mx::MxArray) = jsparse(mx)
+convert(::Type{Dict}, mx::MxArray) = jdict(mx)
```
Comment (yuyichao) [inline src/mxarray.jl]: And all of these also needs deprecations.

**Comment by musm** [comment]
@yuyichao and @simonster  these changes are just ported over from @twadleigh  at mexjulia.

**Comment by yuyichao** [comment]
That's a very bad argument for why it's the right change.

File: src/mxarray.jl (hunk 2 above)
Comment (musm) [inline src/mxarray.jl]: Good point. I added them here so that the  errors can be triggered at the lowest levels, since these functions are called with `size` , `length` etc.

Where should it be an `unsafe_convert`?

Comment (yuyichao) [inline src/mxarray.jl]: Read the doc at http://julia.readthedocs.io/en/latest/manual/calling-c-and-fortran-code/#auto-conversion

Example of definition at https://github.com/yuyichao/LibALPM.jl/blob/c5b34d781328e64a31f26bbb249563f1cbf8428a/src/handle.jl#L178 and its use at https://github.com/yuyichao/LibALPM.jl/blob/c5b34d781328e64a31f26bbb249563f1cbf8428a/src/handle.jl#L186

Comment (yuyichao) [inline src/mxarray.jl]: Note that I strongly recommend against doing that in this PR.

Comment (musm) [inline src/mxarray.jl]: Thanks, I will open a different PR, to fix this segfault.

File: src/mxarray.jl [hunk 7]
```diff
@@ -652,15 +647,22 @@ function jdict(mx::MxArray)
...
+            Ptr{Void}, (Ptr{Void}, mwIndex, Cint), mx.ptr, 0, i-1)
         fx = MxArray(pv, false)
-        fvals[i] = jvariable(fx)
+        fvals[i] = Any(fx)
     end
     Dict(zip(fnames, fvals))
 end
 
-function jvariable(mx::MxArray)
+function Function(mx::MxArray)
```
Comment (twadleigh) [inline src/mxarray.jl]: I added this in `mexjulia` so that MATLAB function handles could be passed to julia and used as callbacks without any special extra marshaling. However, for this definition to do anything meaningful, `MxArray` needs to be made callable. In the context of embedding Julia in MATLAB, this can be achieved via an underlying invocation of `mexCallMATLABWithTrap`, which I provide in `mexjulia`. I very briefly looked into whether there would be a way to do this in the context of interacting with MATLAB via the engine interface, as `MATLAB.jl` does, but didn't immediately see any obvious analogy. In fact, that was the primary motivation for me to fork the data marshaling code from `MATLAB.jl` into `mexjulia`.
